File: rust/src/206.reverse-linked-list.rs
use crate::Solution;

#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
    pub val: i32,
    pub next: Option<Box<ListNode>>,
}

#[allow(dead_code)]
impl ListNode {
    #[inline]
    fn new(val: i32) -> Self {
        ListNode { next: None, val }
    }
}

#[allow(dead_code)]
impl Solution {
    pub fn reverse_list(head: Option<Box<ListNode>>) -> Option<Box<ListNode>> {
        Self::recursive(head, None)
    }

    fn recursive(
        head: Option<Box<ListNode>>,
        prev: Option<Box<ListNode>>,
    ) -> Option<Box<ListNode>> {
        match head {
            None => prev,
            Some(mut current) => {
                let next = current.next.take();
                current.next = prev;
                Self::recursive(next, Some(current))
            }
        }
    }
}

File: ruby/206.reverse-linked-list.rb
# frozen_string_literal: true

# @param {ListNode} head
# @return {ListNode}
def reverse_list(head)
  recursive_reverse_list(head)
end

def recursive_reverse_list(current, prev = nil)
  return prev unless current

  next_node = current.next
  current.next = prev
  recursive_reverse_list(next_node, current)
end
